Fix crash when clicking on "Interpolation Mode" with nonexistent node path

RealMadvicius opened the pull request:
#81769 AnimationPlayer: Editor crashes when clicking on "Interpolation Mode" with nonexistent node path

- adding a nullptr check on a Node pointer obtained from get_node(NodePath) in case it is null now we wont execute the next instruction

> ClassDB::get_property_info(nd->get_class(), prop, &prop_info);

Which then prevents the crash and leave the PropertyInfo prop_info blank default, so when we reach next instruction

> bool is_angle = prop_info.type == Variant::FLOAT && prop_info.hint_string.find("radians") != -1;

we will have prop_info.type = NIL and
prop_info.hint_string.find("radians") will return -1 since it is empty

So is_angle will be set to false and the code can continue to execute normaly without any crash

* *Bugsquad edit, fixes: #81769*

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

RealMadvicius commented:
Normaly the typo is fixed to respect the clang format, lets cross fingers

AThousandShips commented on editor/animation_track_editor.cpp:
```diff
@@ -2698,7 +2698,11 @@ void AnimationTrackEdit::gui_input(const Ref<InputEvent> &p_event) {
 						StringName prop = npath.get_concatenated_subnames();
 						PropertyInfo prop_info;
-						ClassDB::get_property_info(nd->get_class(), prop, &prop_info);
+
```
This should be inside an if statement instead, like so:
```cpp
						if (nd) {
							StringName prop = npath.get_concatenated_subnames();
							PropertyInfo prop_info;
							ClassDB::get_property_info(nd->get_class(), prop, &prop_info);

							bool is_angle = prop_info.type == Variant::FLOAT && prop_info.hint_string.find("radians") != -1;
							if (is_angle) {
								menu->add_icon_item(get_editor_theme_icon(SNAME("InterpLinearAngle")), TTR("Linear Angle"), MENU_INTERPOLATION_LINEAR_ANGLE);
								menu->add_icon_item(get_editor_theme_icon(SNAME("InterpCubicAngle")), TTR("Cubic Angle"), MENU_INTERPOLATION_CUBIC_ANGLE);
							}
						}
```
Starting from the line below `Node *nd = ap->get_node(ap->get_root())->get_node(NodePath(npath.get_concatenated_names()));`

RealMadvicius replied:
I have ammended my commit with the recommendations you provided

AThousandShips replied:
Looks good! But remove this line

RealMadvicius replied:
Done

lyuma commented on editor/animation_track_editor.cpp:
```diff
@@ -2696,13 +2696,15 @@ void AnimationTrackEdit::gui_input(const Ref<InputEvent> &p_event) {
 					if (ap) {
 						NodePath npath = animation->track_get_path(track);
 						Node *nd = ap->get_node(ap->get_root())->get_node(NodePath(npath.get_concatenated_names()));
```
We should store the result of `ap->get_node(ap->get_root())` in a tempoary `Node*` and also perform a null check on this part.

I checked and this is the only place in the file that performs either of these `get_node` calls so it's probably best to be safe. Not all AnimationPlayer nodes have their root set to `..`

RealMadvicius replied:
I have added the extra check and a little comment in the code to explain why we need it

lyuma reviewed (approved):
I think it makes a lot of sense to do this check. This code generally doesn't touch nodes within the scene

This is the only place in the code that does either call to `get_node()` which is a big warning flag, so we need to be extra safe in this part of the code.

Your change is good, and I propose one additional check to be extra safe on top of what you already did.

RealMadvicius committed:
Fix crash when clicking on "Interpolation Mode" with nonexistent node path

issue reference [https://github.com/godotengine/godot/issues/81769]
#81769 AnimationPlayer: Editor crashes when clicking on "Interpolation Mode" with nonexistent node path

- adding a nullptr check on a Node pointer obtained from get_node(NodePath) in case it is null now we wont execute the next instruction

> ClassDB::get_property_info(nd->get_class(), prop, &prop_info);

Which then prevents the crash

AThousandShips commented on editor/animation_track_editor.cpp:
```diff
@@ -2695,18 +2695,25 @@ void AnimationTrackEdit::gui_input(const Ref<InputEvent> &p_event) {
-						StringName prop = npath.get_concatenated_subnames();
-						PropertyInfo prop_info;
-						ClassDB::get_property_info(nd->get_class(), prop, &prop_info);
+						Node *aApRootNode = ap->get_node(ap->get_root_node());
```
```suggestion
						Node *a_ap_root_node = ap->get_node(ap->get_root_node());
```
Use `snake_case`

RealMadvicius replied:
my bad I'm too used to camel case from my workplace, I'm gonna fix this

AThousandShips commented on editor/animation_track_editor.cpp:
```diff
@@ -2695,18 +2695,25 @@ void AnimationTrackEdit::gui_input(const Ref<InputEvent> &p_event) {
+						Node *aApRootNode = ap->get_node(ap->get_root_node());
+						Node *nd = nullptr;
+						//we must test that we have a valid ApRootNode before trying to access its content to init the nd Node
+						if(aApRootNode) {
```
```suggestion
						if (a_ap_root_node) {
```

AThousandShips commented on editor/animation_track_editor.cpp:
```diff
@@ -2695,18 +2695,25 @@ void AnimationTrackEdit::gui_input(const Ref<InputEvent> &p_event) {
-						ClassDB::get_property_info(nd->get_class(), prop, &prop_info);
+						Node *aApRootNode = ap->get_node(ap->get_root_node());
+						Node *nd = nullptr;
+						//we must test that we have a valid ApRootNode before trying to access its content to init the nd Node
```
```suggestion
						// We must test that we have a valid ApRootNode before trying to access its content to init the nd Node.
```

AThousandShips commented:
Please update your commit title to be more descriptive, the title of the PR is good

RealMadvicius commented:
I have edited the commit title to correspond to the PR title as asked

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:

harrisyu commented:
With this commit, we found a bug, after you moved a node, animationplayer lost it's path. does it execute when node in / out of tree?

YuriSizov commented:
@harrisyu Please report your bug as a new issue with a reproduction project attached.
